File: 0045-jump-game-ii/0045-jump-game-ii.cpp
class Solution {
public:
    int solve(int i,vector<int>& nums,int n,vector<int> &dp){
        if(i==n-1){
            return 0;
        }
        if(i>=n){
            dp[i]=INT_MAX;
        }
        if(dp[i]!=-1){
            return dp[i];
        }
        int steps=INT_MAX;
        for(int j=1;j<=nums[i] && i + j < n ;j++){
            int subProb=solve(j+i,nums,n,dp);
            if(subProb!=INT_MAX){
                steps=min(steps,subProb+1);
            }
        }
        return dp[i]=steps;
    }
    int jump(vector<int>& nums) {
        vector<int> dp(nums.size(),-1);
        return solve(0,nums,nums.size(),dp);
    }
};

File: 0045-jump-game-ii/0045-jump-game-ii.java
class Solution {
    public int jump(int[] nums) {
        int l = 0, r = 0, j = 0;
        while (r < nums.length - 1) {
            int far = 0;
            for (int i = l; i <= r; i++) {
                far = Math.max(far, i + nums[i]);
            }
            l = r + 1;
            r = far;
            j++;
        }
        return j;
    }
}